src/slice.c

```diff
@@ -207,34 +207,39 @@ r_obj* df_slice(r_obj* x, r_obj* subscript) {
   return out;
 }
 
+static
+r_obj* bracket_dispatch(r_obj* x, r_obj* subscript) {
+  return vctrs_dispatch2(
+    syms_bracket, fns_bracket,
+    syms_x, x,
+    syms_i, subscript
+  );
+}
+static
+r_obj* bracket_shaped_dispatch(r_obj* x, r_obj* subscript) {
+  return vctrs_dispatch2(
+    syms.bracket_shaped_dispatch, fns.bracket_shaped_dispatch,
+    syms_x, x,
+    syms_i, subscript
+  );
+}
 
 r_obj* vec_slice_fallback(r_obj* x, r_obj* subscript) {
```

Comment:
The point of `vec_slice_fallback()` is now to completely handle the fallback slice case - which is basically how we have been trying to use it this whole time.

It first hands off to `[` and then optionally calls `vec_restore()` if it looks like we needed to. This allows us to localize the "maybe restore" logic into one place, so we don't forget to do it whenever we use `vec_slice_fallback()` elsewhere.